spring-kafka/src/main/java/org/springframework/kafka/ReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,304 @@
+		}
+	}
+
+	@Override
+	public synchronized void start() {
+		try {
+			afterPropertiesSet();
+		}
+		catch (Exception e) {
+			throw new KafkaException("Failed to initialize", e);
+		}
+		this.replyContainer.start();
+		this.running = true;
+	}
+
+	@Override
+	public synchronized void stop() {
+		this.running = false;
+		this.replyContainer.stop();
+	}
+
+	@Override
+	public synchronized boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public RequestReplyFuture<K, V, R> sendAndReceive(ProducerRecord<K, V> record) {
+		CorrelationKey correlationId = createCorrelationId();
+		record.headers().add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.correlationId));
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("Sending: " + record + " with correlationId: " + correlationId);
+		}
+		TemplateRequestReplyFuture<K, V, R> future = new TemplateRequestReplyFuture<>();
+		this.futures.put(correlationId, future);
+		future.setSendFuture(send(record));
+		this.scheduler.schedule(() -> {
+			RequestReplyFuture<K, V, R> removed = this.futures.remove(correlationId);
+			if (removed != null) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("Reply timed out for: " + record + " with correlationId: " + correlationId);
+				}
+				removed.setException(new KafkaException("Reply timed out"));
+			}
+		}, Instant.now().plusMillis(this.replyTimeout));
+		return future;
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		if (!this.schedulerSet) {
+			((ThreadPoolTaskScheduler) this.scheduler).destroy();
+		}
+	}
+
+	/**
+	 * Subclasses can override this to generate custom correlation ids.
+	 * The default implementation is a 16 byte representation of a UUID.
+	 * @return the key.
+	 */
+	protected CorrelationKey createCorrelationId() {
+		UUID uuid = UUID.randomUUID();
+		byte[] bytes = new byte[16];
+		ByteBuffer bb = ByteBuffer.wrap(bytes);
+		bb.putLong(uuid.getMostSignificantBits());
+		bb.putLong(uuid.getLeastSignificantBits());
+		return new CorrelationKey(bytes);
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, R>> data) {
+		data.forEach(record -> {
+			Iterator<Header> iterator = record.headers().iterator();
+			CorrelationKey correlationId = null;
+			while (correlationId == null && iterator.hasNext()) {
+				Header next = iterator.next();
+				if (next.key().equals(KafkaHeaders.CORRELATION_ID)) {
+					correlationId = new CorrelationKey(next.value());
+				}
+			}
+			if (correlationId == null) {
+				this.logger.error("No correlationId found in reply: " + record);
```

Comment:
I think this not enough. Not found. And? What now ? Did we do anything wrong?
I mean let's add some explanation into the log message what is wrong in this case.